Tweek.JPad/Rule.fs

```diff
@@ -199,20 +203,31 @@ module Matcher =
                     | BinaryOp.CompareOp compare_op, _ -> (|>) prefix >> evaluateComparison comaprer compare_op op_value
                     | BinaryOp.TimeOp time_op, _ -> evaluateTimeComparison prefix time_op op_value
                     | BinaryOp.StringOp string_op, JsonValue.String string_value -> (|>) prefix >> Option.map JsonExtensions.AsString >> evaluateStringComparison string_op string_value
-                    | BinaryOp.ContainsOp, contains_value -> (|>) prefix >> (evaluateContains contains_value |> falseOnNone) 
-
+                    | BinaryOp.ContainsOp, contains_value -> (|>) prefix >> (evaluateContains contains_value |> falseOnNone)
+                    | BinaryOp.AnyOp, any_value -> (|>) prefix >> (evaluateAny comparers any_value |> falseOnNone)
+                    | BinaryOp.AllOp, all_value -> (|>) prefix >> (evaluateAll comparers all_value |> falseOnNone)
                     | _ -> raise (ParseError "invalid binary matcher")
                 | Empty -> (fun context->true)
-
-        
         CompileExpression "" exp
-
-    let parse (schema:JsonValue) = parsePropertySchema ConjuctionOp.And ComparisonType.Auto schema 
-
-    let createEvaluator (settings: ParserSettings) (matcher: MatcherExpression) =
-        let comparersMaps = (settings.Comparers |> toMap).Add("date", new ComparerDelegate(fun x -> DateTime.Parse(x) :> IComparable))
-        (matcher |> (compile_internal comparersMaps))
-
+        
+    and private evaluateAny (comparers:Map<string,ComparerDelegate>) (leftValue:ComparisonValue) =
+        let context = fun (value: JsonValue) (name: string) -> match name with
+                                                               | "" -> Some value
+                                                               | _ -> value.TryGetProperty(name)
+        let eval = parse leftValue |> createEvaluator comparers
+        fun (rightValue:JsonValue) -> match rightValue with
+                                       | Array arr -> Array.exists (fun x-> (context(x)) |> eval ) arr
+                                       | _ -> false
+
+    and private evaluateAll (comparers:Map<string,ComparerDelegate>) (leftValue:ComparisonValue) =
+        let context = fun (value: JsonValue) (name: string) -> match name with
```

Comment:
Add utility function (no need to define it in the function scope and it'll reduce duplication)
```fsharp
let private createNestedContext (value: JsonValue) (name: string) = match name with
                                                                           | "" -> Some value
                                                                           | _ -> value.TryGetProperty(name)
```
(Can be added in the start of the file.)

also ```(context(x)) |> eval``` can be written as:
```x |> createNestedContext |> eval```
which is more ideomatic F# and has less parens

Reply:
Or even cleaner:
```Array.exists (createNestedContext >> eval)```

Reply:
Also, lookup property matchers in Tweek are case insensitive, so to preserve this behavior we should use context-mapping that isn't case-sensitive:
```
value.Properties() |> Array.tryFind (fun (p, _)->StringComparer.OrdinalIgnoreCase.Compare(p, name) = 0) |> Option.map snd
``` instead of just ```value.TryGetProperty(name)

This case should probably be tested.